### PR TITLE
Feature: test for valid/invalid promises

### PR DESCRIPTION
**Pull request by maxkorz**
Changes in this pull request:
1. Refactored promiseBtn setter. An isPromise check was added.
2. Added tests to check that all invalid values (not promises) handled correctly
3. Added tests to check that all valid values (see below) handled correctly

List of valid values:
1. native promise
2. jQuery Deferred
3. jQuery Deferred Promise
4. Bluebird Promise
5. RxJs Observable

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -80,7 +84,7 @@
     "rxjs": "^5.1.0",
     "ts-node": "~2.0.0",
     "tslint": "~4.5.0",
-    "typescript": "~2.2.0",
+    "typescript": "~2.3.0",
```
Comment (maxkorz) [inline package.json]: I had to update typescript to next minor version because bluebird wasn't compiling.

File: src/promise-btn.directive.ts [hunk 2]
```diff
@@ -33,12 +33,21 @@ export class PromiseBtnDirective implements OnDestroy, AfterContentInit {
...
-      this.promise = promise.toPromise();
-    } else {
-      this.promise = promise;
+  set promiseBtn(request: any) {
+    const isObservable: boolean = request instanceof Observable;
+    const isPromise: boolean = request instanceof Promise || (
+      request !== null &&
+      typeof request === 'object' &&
+      typeof request.then === 'function' &&
+      typeof request.catch === 'function'
```
Comment (maxkorz) [inline src/promise-btn.directive.ts]: this check makes sense as in initPromiseHandler we call both then and catch functions
```javascript
      promise
        .then(resolveLoadingState)
        .catch(resolveLoadingState);
```

File: src/promise-btn.directive.ts [hunk 3]
```diff
@@ -33,12 +33,21 @@ export class PromiseBtnDirective implements OnDestroy, AfterContentInit {
...
+    );
+
+    if (isObservable) {
+      this.promise = request.toPromise();
+    } else if (isPromise) {
+      this.promise = request;
```
Comment (maxkorz) [inline src/promise-btn.directive.ts]: if passed value is not Observable and is not Promise, `this.promise` won't be set. It allows us to simplify checks below:
` if (!this.promise || !this.promise.then) {`
becomes
` if (!this.promise) {`

File: src/promise-btn.directive.ts [hunk 4]
```diff
@@ -137,15 +147,11 @@ export class PromiseBtnDirective implements OnDestroy, AfterContentInit {
...
    * this.cfg.minDuration into account if given.
-   * @param {Object}promise
    * @param {Object}btnEl
    */
 
-  initPromiseHandler(promise: any, btnEl: HTMLElement) {
-    // return if something else then a promise is passed
-    if (!promise || !promise.then) {
-      return;
-    }
```
Comment (maxkorz) [inline src/promise-btn.directive.ts]: Only `checkAndInitPromiseHandler` calls `initPromiseHandler`, and, as we can see the `this.promise` variable was already checked in `checkAndInitPromiseHandler`, so check in `initPromiseHandler` is redundant and can be deleted.

**Comment by johannesjo** [comment]
Thanks again for your work! Looks great! 

I like a lot how you explain what you're doing. Personally I would prefer to have the comments made via the github interface to be directly in the code, so the next person working with it has a better indication on why things are done the way they are. Would you mind copying that over to the code?

**Comment by maxkorz** [comment]
@johannesjo thanks! Of course, you can copy my comments to the code.

**Pull request merged**